Build fails due to the use of deprecated C++ feature std::unary_function

### Description

```
/wrkdirs/usr/ports/science/py-hoomd-blue/work/hoomd-blue-4.0.1/hoomd/ParticleData.cc:2893:32: error: no template named 'unary_function' in namespace 'std'; did you mean '__unary_function'?
struct comm_flag_select : std::unary_function<const unsigned int, bool>
                          ~~~~~^~~~~~~~~~~~~~
                               __unary_function
/usr/include/c++/v1/__functional/unary_function.h:46:1: note: '__unary_function' declared here
using __unary_function = __unary_function_keep_layout_base<_Arg, _Result>;
^
1 error generated.
```

clang-16 is more strict than the previous versions and it now fails when deprecated C++ features are used.

std::unary_function was removed in C++17, see [here](https://en.cppreference.com/w/cpp/utility/functional/unary_function), and you use C++17.

You need to remove the use of deprecated features and make a release.


Thank you,
Yuri

FreeBSD 13.2

### Script

```python
n/a
```


### Input files

n/a

### Output

```shell
n/a
```


### Expected output

n/a

### Platform

Linux

### Installation method

Compiled from source

### HOOMD-blue version

4.0.1

### Python version

3.9

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -36,7 +36,7 @@ jobs:
     name: Build release tarball
     runs-on: ubuntu-latest
     container:
-      image: glotzerlab/ci:2023.06.19-ubuntu20.04
+      image: glotzerlab/ci:2023.07.24-ubuntu20.04
 
     steps:
       - name: Checkout

--- a/.github/workflows/templates/configurations.yml
+++ b/.github/workflows/templates/configurations.yml
@@ -24,6 +24,8 @@ validate_configurations:
 # Configurations to build and test only rarely, such as just before a release.
 # There should be no overlap between this list and `unit_test_configurations`
 release_test_configurations:
+- config: "[clang16_py311, mpi, llvm]"
+- config: "[clang15_py311, mpi, llvm]"
 - config: "[clang13_py310, llvm]"
 - config: "[clang12_py310, llvm]"
 - config: "[clang11_py310, llvm]"

--- a/.github/workflows/templates/workflow.yml
+++ b/.github/workflows/templates/workflow.yml
@@ -1,5 +1,5 @@
 <% block name %><% endblock %>
-<% set container_prefix="glotzerlab/ci:2023.06.19" %>
+<% set container_prefix="glotzerlab/ci:2023.07.24" %>
 
 <% block concurrency %>
 concurrency:

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -59,7 +59,7 @@ jobs:
     name: Build [${{ join(matrix.config, '_') }}]
     runs-on: ${{ matrix.build_runner }}
     container:
-      image: glotzerlab/ci:2023.06.19-${{ matrix.config[0] }}
+      image: glotzerlab/ci:2023.07.24-${{ matrix.config[0] }}
     strategy:
       matrix:
         include:
@@ -168,7 +168,7 @@ jobs:
     needs: build
     runs-on: ${{ matrix.test_runner }}
     container:
-      image: glotzerlab/ci:2023.06.19-${{ matrix.config[0] }}
+      image: glotzerlab/ci:2023.07.24-${{ matrix.config[0] }}
       options: ${{ matrix.test_docker_options }} -e CUDA_VISIBLE_DEVICES
     strategy:
       matrix:
@@ -229,7 +229,7 @@ jobs:
     needs: build
     runs-on: ${{ matrix.test_runner }}
     container:
-      image: glotzerlab/ci:2023.06.19-${{ matrix.config[0] }}
+      image: glotzerlab/ci:2023.07.24-${{ matrix.config[0] }}
       options: ${{ matrix.test_docker_options }} -e CUDA_VISIBLE_DEVICES
     strategy:
       matrix:
@@ -283,7 +283,7 @@ jobs:
     needs: build
     runs-on: ${{ matrix.test_runner }}
     container:
-      image: glotzerlab/ci:2023.06.19-${{ matrix.config[0] }}
+      image: glotzerlab/ci:2023.07.24-${{ matrix.config[0] }}
       options: ${{ matrix.test_docker_options }} -e CUDA_VISIBLE_DEVICES
     strategy:
       matrix:
@@ -334,10 +334,12 @@ jobs:
     name: Build [${{ join(matrix.config, '_') }}]
     runs-on: ${{ matrix.build_runner }}
     container:
-      image: glotzerlab/ci:2023.06.19-${{ matrix.config[0] }}
+      image: glotzerlab/ci:2023.07.24-${{ matrix.config[0] }}
     strategy:
       matrix:
         include:
+        - {config: [clang16_py311, mpi, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
+        - {config: [clang15_py311, mpi, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
         - {config: [clang13_py310, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
         - {config: [clang12_py310, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
         - {config: [clang11_py310, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
@@ -445,11 +447,13 @@ jobs:
     needs: build_release
     runs-on: ${{ matrix.test_runner }}
     container:
-      image: glotzerlab/ci:2023.06.19-${{ matrix.config[0] }}
+      image: glotzerlab/ci:2023.07.24-${{ matrix.config[0] }}
       options: ${{ matrix.test_docker_options }} -e CUDA_VISIBLE_DEVICES
     strategy:
       matrix:
         include:
+        - {config: [clang16_py311, mpi, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
+        - {config: [clang15_py311, mpi, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
         - {config: [clang13_py310, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
         - {config: [clang12_py310, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
         - {config: [clang11_py310, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
@@ -512,11 +516,13 @@ jobs:
     needs: build_release
     runs-on: ${{ matrix.test_runner }}
     container:
-      image: glotzerlab/ci:2023.06.19-${{ matrix.config[0] }}
+      image: glotzerlab/ci:2023.07.24-${{ matrix.config[0] }}
       options: ${{ matrix.test_docker_options }} -e CUDA_VISIBLE_DEVICES
     strategy:
       matrix:
         include:
+        - {config: [clang16_py311, mpi, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
+        - {config: [clang15_py311, mpi, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
         - {config: [clang13_py310, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
         - {config: [clang12_py310, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }
         - {config: [clang11_py310, llvm], build_runner: [self-hosted,jetstream2,CPU], test_runner: ubuntu-latest, test_docker_options: '' }

